src/Features/Core/Portable/ConvertForToForEach/AbstractConvertForToForEachCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,504 @@
+                {
+                    return;
+                }
+            }
+
+            var collectionType = semanticModel.GetTypeInfo(collectionExpression, cancellationToken);
+            if (collectionType.Type == null && collectionType.Type.TypeKind == TypeKind.Error)
+            {
+                return;
+            }
+
+            var containingType = semanticModel.GetEnclosingNamedType(context.Span.Start, cancellationToken);
+            if (containingType == null)
+            {
+                return;
+            }
+
+            var ienumerableType = semanticModel.Compilation.GetSpecialType(SpecialType.System_Collections_Generic_IEnumerable_T);
+            var ienumeratorType = semanticModel.Compilation.GetSpecialType(SpecialType.System_Collections_Generic_IEnumerator_T);
+
+            // make sure the collection can be iterated.
+            if (!TryGetIterationElementType(
+                    containingType, collectionType.Type,
+                    ienumerableType, ienumeratorType,
+                    out var iterationType))
+            {
+                return;
+            }
+
+            // If the user uses the iteration variable for any other reason, we can't convert this.
+            var bodyStatements = GetBodyStatements(forStatement);
+            foreach (var statement in bodyStatements)
+            {
+                if (iterationVariableIsUsedForMoreThanCollectionIndex(statement))
+                {
+                    return;
+                }
+            }
+
+            // Looks good.  We can convert this.
+            context.RegisterRefactoring(new MyCodeAction(GetTitle(),
+                c => ConvertForToForEachAsync(
+                    document, forStatement, iterationVariable, collectionExpression,
+                    containingType, collectionType.Type, iterationType, c)));
+
+            // local functions
+            bool iterationVariableIsUsedForMoreThanCollectionIndex(SyntaxNode current)
+            {
+                if (syntaxFacts.IsIdentifierName(current))
+                {
+                    syntaxFacts.GetNameAndArityOfSimpleName(current, out var name, out _);
+                    if (name == iterationVariable.ValueText)
+                    {
+                        // found a reference.  make sure it's only used inside something like
+                        // list[i]
+
+                        if (!syntaxFacts.IsSimpleArgument(current.Parent) ||
+                            !syntaxFacts.IsElementAccessExpression(current.Parent.Parent.Parent))
+                        {
+                            // used in something other than accessing into a collection.
+                            // can't convert this for-loop.
+                            return true;
+                        }
+
+                        var arguments = syntaxFacts.GetArgumentsOfArgumentList(current.Parent.Parent);
+                        if (arguments.Count != 1)
```

Comment:
Will this handle indexers with more than one parameters, where all but the first are optional?

```
    public T this[int i, int j = 0]
    {
        get { return arr[i, j]; }
        set { arr[i, j] = value; }
    }
```

Reply:
Yes.  it should.  In that case, allowing the conversion seems appropriate.